Assets/Scripts/Internal/DaggerfallSky.cs

```diff
@@ -400,31 +388,189 @@ private void LoadCurrentSky(int targetFrame)
+            if (skyIndex >= 0 && skyIndex <= 7)
+                vanillaNightSky = 3;
+            else if (skyIndex >= 8 && skyIndex <= 15)
+                vanillaNightSky = 1;
+            else if (skyIndex >= 16 && skyIndex <= 23)
+                vanillaNightSky = 2;
+            else
+                vanillaNightSky = 0;
+
+            //TYPE 1: NIGHTIME VANILLA
+            string baseName = string.Format("NITE{0:00}I0.IMG", vanillaNightSky);
+            TextureReplacement.TryImportTexture(baseName, false, out westTexture);
+
+            //Vanilla worked!
+            if (westTexture != null)
+            {
+                //Note east is copied from west
+                eastTexture = westTexture;
+                return true;
+            }
+
+            //TYPE 2: FULL NIGHTIME
+            baseName = string.Format("NITEFULL{0:00}I0.IMG", skyIndex);
+            TextureReplacement.TryImportTexture(baseName, false, out westTexture);
+
+            //Full worked!
+            if (westTexture != null)
+            {
+                //Note east is copied from west
+                eastTexture = westTexture;
+                return true;
+            }
+
+            //TYPE 3: NIGHTIME VANILLA
+            baseName = string.Format("NITE{0:00}I0.IMG", vanillaNightSky);
+            string eastName = Path.GetFileNameWithoutExtension(baseName) + "-0" + Path.GetExtension(baseName);
+            string westName = Path.GetFileNameWithoutExtension(baseName) + "-1" + Path.GetExtension(baseName);
+
+            TextureReplacement.TryImportTexture(eastName, false, out eastTexture);
+            TextureReplacement.TryImportTexture(westName, false, out westTexture);
+
+            //Vanilla with east+west worked!
+            if (westTexture != null && eastTexture != null)
+                return true;
+
+            //TYPE 2: FULL NIGHTIME
+            baseName = string.Format("NITEFULL{0:00}I0.IMG", skyIndex);
+            eastName = Path.GetFileNameWithoutExtension(baseName) + "-0" + Path.GetExtension(baseName);
+            westName = Path.GetFileNameWithoutExtension(baseName) + "-1" + Path.GetExtension(baseName);
+
+            TextureReplacement.TryImportTexture(eastName, false, out eastTexture);
+            TextureReplacement.TryImportTexture(westName, false, out westTexture);
+
+            //Full with east+west worked!
+            if (westTexture != null && eastTexture != null)
+                return true;
 
+            westTexture = null;
+            eastTexture = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Loads day sky from arena files
+        /// </summary>
+        /// <param name="skyIndex"></param>
+        /// <returns>SkyColors loaded</returns>
+        private SkyColors LoadVanillaDaySky(int skyIndex, int frame)
+        {
+            SkyFile skyFile = new SkyFile(Path.Combine(dfUnity.Arena2Path, SkyFile.IndexToFileName(skyIndex)), FileUsage.UseMemory, true);
```

Comment:
I would like for SkyFile to be loaded and cached only when the sky file itself changes, rather than loading and discarding away the entire SkyFile every frame. This is why the SkyFile data is currently handled the way it is in this class. This is particularly noticeable with increased timescale (e.g. set_timescale 1000), it generates large GC spikes and more disk accesses that could be avoided.

Reply:
From a quick glance, I'm surprised it runs EVERY frame, since there is a check on line 163 that checks if the skyindex is different. regardless, I see what you mean about there lacking any caching of the skyfile, I agree that this should be fixed in the future.

Reply:
Yep, I just mean each time the sky frame changes (not each frame rendered).

My mental map was that it cached the SkyFile and only loaded a new SkyFile when the sky file changed - but doesn't look like I implemented that way for some reason. Not sure why my memory doesn't mesh up here. Maybe it was something I meant to do back in 2014/15 when I first wrote this and just never got back to it.

Reply:
One way to do it would be instead of caching the skyfile itself, it could simply cache the textures in a dictionary, with the key being the filename.

Reply:
I'd rather not cache each Texture2D. These are substantially larger than the compact SkyFile format, and modded textures could be larger still. I mainly just want to avoid hitting the disk for the entire SkyFile every time the sky frame changes, as most of the time it will be the same SkyFile as previous frame. I also bet disk caching makes that a pretty minimal problem anyway, so the whole concern is probably misplaced on my end. :)

I'll loop back to this one later. I just made an incorrect assessment based on faulty memory when looking at your changes without properly comparing to original. So much time has passed since this class was created that I should have spent more time re-familiarising myself with my own work first. I apologise, and will try to avoid this in future.